src/app/loadout-builder/processWorker/process.ts

```diff
@@ -0,0 +1,494 @@
+              // This is an awkward implementation to save garbage allocations.
+              let tiers = '';
+              let totalTier = 0;
+              let index = 1;
+              for (const statKey in stats) {
+                tiers += statTier(stats[statKey]);
+                totalTier += statTier(stats[statKey]);
+                if (index < statOrder.length) {
+                  tiers += ',';
+                }
+                index++;
+              }
+
+              // While we have less than RETURNED_ARMOR_SETS sets keep adding and keep track of the lowest total tier.
+              if (totalTier < lowestTier) {
+                if (setCount <= RETURNED_ARMOR_SETS) {
+                  lowestTier = totalTier;
+                } else {
+                  continue;
+                }
+              }
+
+              let groupedSetsInTotalTier = groupedSets[totalTier];
+
+              if (!groupedSetsInTotalTier) {
+                groupedSets[totalTier] = {};
+                groupedSetsInTotalTier = groupedSets[totalTier];
+              }
+
+              const existingSetAtTier = groupedSetsInTotalTier[tiers];
+              if (existingSetAtTier) {
+                existingSetAtTier.sets.push({
+                  armor: armor.map((items) => items.map((item) => item.id)),
+                  statChoices,
+                  maxPower,
+                });
+                if (maxPower > existingSetAtTier.maxPower) {
+                  existingSetAtTier.firstValidSet = firstValidSet.map((item) => item.id);
+                  existingSetAtTier.maxPower = maxPower;
+                  existingSetAtTier.firstValidSetStatChoices = statChoices;
+                }
+              } else {
+                // First of its kind
+                groupedSetsInTotalTier[tiers] = {
+                  sets: [
+                    {
+                      armor: armor.map((items) => items.map((item) => item.id)),
+                      statChoices,
+                      maxPower,
+                    },
+                  ],
+                  stats: stats as {
+                    [statType in StatTypes]: number;
+                  },
+                  // TODO: defer calculating first valid set / statchoices / maxpower?
+                  firstValidSet: firstValidSet.map((item) => item.id),
+                  firstValidSetStatChoices: statChoices,
+                  maxPower,
+                };
+              }
+
+              setCount++;
+
+              if (setCount > RETURNED_ARMOR_SETS && groupedSets[lowestTier]) {
+                const lowestTierGroup = groupedSets[lowestTier];
+                const lowestTierStatMixes = Object.keys(lowestTierGroup);
+
+                if (lowestTierStatMixes.length) {
+                  // Stats are sorted by order so this perfers sets by stat order preference
+                  const sortedStatMixes = lowestTierStatMixes.sort((a, b) => b.localeCompare(a));
```

Comment:
This is pretty rough to have in the hot loop - I wonder if we could use an insertion sort or a minheap to make sure we always keep things in sorted order so we don't need to do this.

Reply:
Yeah I can look into that, a minheap would be nice but would we want to use a library or is there something custom?

Another option might be to do `lowestTierStatMixes.sort().reverse()` as from what I understand of JS it will use the underlying c++ directly without the need to run the compare function. This would potentially be quicker but I wasn't super fussed while doing it as I know it is a relatively small array.